### PR TITLE
Feat: Buddy List and Music Compose

### PR DESCRIPTION
**Pull request by Zintixx**
## Buddy List
Implemented Buddy List. This also includes partial Block functionality. 
Buddy system works by a shared ID system. Both players are connected by the ID in order to get each other's information on each Buddy List.
There's a few things in the current implementation that are temporary (as indicated as comments) until a database is set up.

## Music Compose
Adjusted the WriteItem packet to work with UGC Scores.
Parsing "isCustomNote" to find the items that are UGCScores.
Music composing should now work as intended. 
Listening to music for _other_ players in the map still does not work. I left a comment as to what needs to be fixed.

### Etc
Moved all of the function Item data to it's own class. This should make it a bit cleaner.

### CODE REVIEW

File: Maple2Storage/Types/Metadata/ItemMetadata.cs [hunk 1]
```diff
@@ -223,6 +213,85 @@ public ItemBreakReward(int id, int count)
         public override string ToString() => $"Id: {Id}, Amount: {Count}";
     }
 
+    [XmlType]
+    public class Function
```
Comment (Sparkymod) [inline Maple2Storage/Types/Metadata/ItemMetadata.cs:217]: `Function` probably would be confuse with something else, perhaps `ItemFunction` or `ItemMetadataFunction`

File: MapleServer2/PacketHandlers/Game/BuddyHandler.cs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+                case BuddyMode.CancelRequest:
+                    HandleCancelRequest(session, packet);
+                    break;
+                default:
+                    IPacketHandler<GameSession>.LogUnknownMode(mode);
+                    break;
+            }
+        }
+
+        public enum BuddyNotice : byte
```
Comment (Sparkymod) [inline MapleServer2/PacketHandlers/Game/BuddyHandler.cs:66]: Can move this enum right below the other enum on line 28

File: MapleServer2/PacketHandlers/Game/BuddyHandler.cs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+            session.Send(BuddyPacket.UpdateBuddy(buddy));
+            session.Send(BuddyPacket.LoginLogoutNotifcation(buddy));
+
+            if (buddy.Friend != null) // TODO: Change to send if online
+            {
+                buddy.Friend.Session.Send(BuddyPacket.UpdateBuddy(buddyFriend));
+                buddy.Friend.Session.Send(BuddyPacket.AcceptNotification(buddyFriend));
+            }
+        }
+        private static void HandleDecline(GameSession session, PacketReader packet)
```
Comment (Sparkymod) [inline MapleServer2/PacketHandlers/Game/BuddyHandler.cs:195]: Empty space between methods

File: MapleServer2/PacketHandlers/Game/InstrumentHandler.cs [hunk 4]
```diff
@@ -121,6 +125,30 @@ private static void HandleStopScore(GameSession session/*, PacketReader packet*/
...
+            item.Score.Length = length;
+            item.Score.Type = instrumentType;
+            item.Score.Title = scoreName;
+            item.Score.Composer = session.Player.Name;
+            item.Score.ComposerCharacterId = session.Player.CharacterId;
+            item.Score.Notes = scoreNotes;
+
+            session.Send(InstrumentPacket.Compose(item));
+        }
         private static void HandleFireworks(GameSession session)
```
Comment (Sparkymod) [inline MapleServer2/PacketHandlers/Game/InstrumentHandler.cs:152]: Empty line between methods

File: MapleServer2/Packets/BuddyPacket.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        public static Packet BlockNotification(string playerName)
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.BUDDY);
+            pWriter.WriteEnum(BuddyPacketMode.BlockNotification);
+            pWriter.WriteByte();
+            pWriter.WriteUnicodeString(playerName);
+            return pWriter;
+        }
+
+
```
Comment (Sparkymod) [inline MapleServer2/Packets/BuddyPacket.cs:150]: Extra line here

**Review by Sparkymod — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/BuddyHandler.cs [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                return;
+            }
+
+            if (targetPlayer == session.Player)
+            {
+                session.Send(BuddyPacket.Notice((byte) BuddyNotice.CannotAddSelf, targetPlayer.Name));
+                return;
+            }
+
+            if ((byte) session.Player.BuddyList.Count(b => b.Blocked != true) >= 100) // 100 is friend limit
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/BuddyHandler.cs:97]: can this just be `!b.Blocked`?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

**Review by Sparkymod — APPROVED** [review]

File: MapleServer2/Packets/InstrumentPacket.cs [hunk 7]
```diff
@@ -50,19 +53,28 @@ public static Packet StopImprovise(IFieldObject<Player> player)
...
+
+            switch (score.IsCustomScore)
+            {
+                case true:
+                    pWriter.WriteMapleString(score.Score.Notes);
+                    break;
+                case false:
+                    pWriter.WriteUnicodeString(score.FileName);
+                    break;
+            }
```
Comment (angelotadeucci) [inline MapleServer2/Packets/InstrumentPacket.cs:77]:
```suggestion
           if (score.IsCustomScore)
            {
                pWriter.WriteMapleString(score.Score.Notes);
            }
            else
            {
                pWriter.WriteUnicodeString(score.FileName);
            }
```

File: MapleServer2/Packets/BuddyPacket.cs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        {
+            pWriter.WriteLong(buddy.SharedId);
+            pWriter.WriteLong(buddy.Friend.CharacterId);
+            pWriter.WriteLong(buddy.Friend.AccountId);
+            pWriter.WriteUnicodeString(buddy.Friend.Name);
+            pWriter.WriteUnicodeString(buddy.Message);
+            pWriter.WriteShort();
+            pWriter.WriteInt(); // player Home map ID
+            pWriter.WriteInt((int) buddy.Friend.Job);
+            pWriter.WriteInt((int) buddy.Friend.JobCode);
```
Comment (angelotadeucci) [inline MapleServer2/Packets/BuddyPacket.cs:194]:
```suggestion
            pWriter.WriteEnum(buddy.Friend.Job);
            pWriter.WriteEnum(buddy.Friend.JobCode);
```

File: MapleServer2/Packets/BuddyPacket.cs [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            pWriter.WriteUnicodeString(buddy.Friend.Motto);
+            pWriter.WriteUnicodeString(buddy.BlockReason);
+            pWriter.WriteInt();
+            pWriter.WriteInt();
+            pWriter.WriteInt();
+            pWriter.WriteUnicodeString(buddy.Friend.HomeName);
+            pWriter.WriteLong();
+            pWriter.WriteInt(buddy.Friend.Trophy[0]);
+            pWriter.WriteInt(buddy.Friend.Trophy[1]);
+            pWriter.WriteInt(buddy.Friend.Trophy[2]);
```
Comment (angelotadeucci) [inline MapleServer2/Packets/BuddyPacket.cs:212]: Up to you.

```suggestion
           foreach (int trophyCount in buddy.Friend.Trophy)
            {
                pWriter.WriteInt(trophyCount);
            }
```

**Review by angelotadeucci — CHANGES_REQUESTED** [review]

**Review by angelotadeucci — APPROVED** [review]

**Review by AlanMorel — APPROVED** [review]

**Review by angelotadeucci — APPROVED** [review]

**Pull request merged**